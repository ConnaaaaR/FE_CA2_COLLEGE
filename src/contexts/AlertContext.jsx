import React, { createContext, useState, useContext } from "react";

const AlertContext = createContext();

export const useAlert = () => useContext(AlertContext);

export const AlertProvider = ({ children }) => {
	const [alert, setAlert] = useState({ isOpen: false, type: "", message: "" });

	const showAlert = (type, message) => {
		setAlert({ isOpen: true, type, message });
		setTimeout(() => setAlert({ isOpen: false, type: "", message: "" }), 15000);
	};

	const closeAlert = () => {
		setAlert({ isOpen: false, type: "", message: "" });
	};

	return (
		<AlertContext.Provider value={{ alert, showAlert, closeAlert }}>
			{children}
		</AlertContext.Provider>
	);
};
