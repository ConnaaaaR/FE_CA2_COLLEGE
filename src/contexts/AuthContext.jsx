import { useContext, createContext, useState } from "react";

const AuthContext = createContext(null);

export function useAuth() {
	const val = useContext(authContext);
	return val;
}

export function AuthProvider(props) {
	const [authenticated, setAuthenticated] = useState(false);

	return (
		<AuthContext.Provider
			value={{
				authenticated,
				onAuthenticated: (auth, token) => {
					setAuthenticated(true);

					if (auth && token) {
						localStorage.setItem("token", token);
					} else if (auth) {
						localStorage.removeItem("token");
					}
				},
			}}
		>
			{props.children}
		</AuthContext.Provider>
	);
}
