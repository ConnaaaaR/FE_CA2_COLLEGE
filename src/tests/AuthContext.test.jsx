import React from "react";
import { render, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { AuthProvider } from "../contexts/AuthContext";

describe("AuthProvider", () => {
	it("renders without crashing", () => {
		render(
			<AuthProvider>
				<div>Test</div>
			</AuthProvider>
		);
	});

	it("sets authenticated state to true if token exists in localStorage", () => {
		const token = "fake_token";
		localStorage.setItem("token", token);

		let component;
		act(() => {
			component = render(
				<AuthProvider>
					<div data-testid="authenticated">Test</div>
				</AuthProvider>
			);
		});

		expect(component.getByTestId("authenticated")).toBeInTheDocument();
	});

	it("sets authenticated state to false if token does not exist in localStorage", () => {
		localStorage.removeItem("token");

		let component;
		act(() => {
			component = render(
				<AuthProvider>
					<div>Test</div>
				</AuthProvider>
			);
		});

		expect(component.container.firstChild).toBeInTheDocument();
		expect(component.container.firstChild).toHaveTextContent("Test");
		expect(component.container.firstChild).toMatchSnapshot();
		expect(component.container.firstChild).not.toHaveAttribute(
			"data-testid",
			"authenticated"
		);
	});

	it("updates authenticated state and localStorage when onAuthenticated is called with auth=true and token", async () => {
		const token = "fake_token";

		let component;
		act(() => {
			component = render(
				<AuthProvider>
					{(props) => {
						const handleOnClick = () => {
							props.onAuthenticated(true, token);
						};
						return (
							<>
								
								<button onClick={handleOnClick}>test</button>;
							</>
						);
					}}
				</AuthProvider>
			);
		});

		const button = component.container.querySelector("button");

		expect(button).toBeInTheDocument();
		expect(button.textContent).toBe("test");

		act(() => {
			button.click();
		});

		expect(localStorage.getItem("token")).toBe(token);
		expect(button).toHaveAttribute("data-testid", "authenticated");
	});

	it("updates authenticated state and removes token from localStorage when onAuthenticated is called with auth=false", () => {
		const token = "fake_token";
		localStorage.setItem("token", token);

		let component;
		act(() => {
			component = render(
				<AuthProvider>
					{({ onAuthenticated }) => (
						<button onClick={() => onAuthenticated(false)}>Logout</button>
					)}
				</AuthProvider>
			);
		});

		const button = component.getByText("Logout");
		expect(button).toBeInTheDocument();

		act(() => {
			component.getByText("Logout").click();
		});

		expect(localStorage.getItem("token")).toBeNull();
		expect(component.container.firstChild).not.toHaveAttribute(
			"data-testid",
			"authenticated"
		);
	});
});
