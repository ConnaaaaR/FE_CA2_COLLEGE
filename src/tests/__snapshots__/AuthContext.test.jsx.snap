// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider sets authenticated state to false if token does not exist in localStorage 1`] = `
<div>
  Test
</div>
`;

exports[`AuthProvider sets authenticated state to true if token exists in localStorage 1`] = `
<div>
  Test
</div>
`;
